File: sdk/tests/contract_vectors.test.ts
/**
 * Cross-language schema contract vector tests (TypeScript/Zod side).
 *
 * These tests load the canonical fixture file `tests/fixtures/contract_vectors.json`
 * and verify that every valid vector parses cleanly via Zod. They also verify that
 * adversarial vectors are rejected, proving divergence detection.
 *
 * ADR reference: docs/adr/ADR-005-schema-contract-enforcement.md
 */
import * as fs from "node:fs";
import * as path from "node:path";
import { describe, it, expect } from "vitest";
import { ZodError } from "zod";
import { RiskScoreSchema } from "../src/schemas";

// ---------------------------------------------------------------------------
// Fixture loading
// ---------------------------------------------------------------------------

type ContractVectors = {
  _contract_version: string;
  risk_score: Record<string, Record<string, unknown>>;
  risk_score_adversarial: Record<string, Record<string, unknown>>;
  trade: Record<string, Record<string, unknown>>;
  asset: Record<string, Record<string, unknown>>;
  required_risk_score_fields: string[];
  required_trade_fields: string[];
  required_asset_fields: string[];
};

function loadFixture(): ContractVectors {
  // Resolve path relative to this test file's location:
  // sdk/tests/contract_vectors.test.ts → ../../tests/fixtures/contract_vectors.json
  const fixturePath = path.resolve(
    __dirname,
    "../../tests/fixtures/contract_vectors.json",
  );
  if (!fs.existsSync(fixturePath)) {
    throw new Error(
      `Contract vectors fixture not found at ${fixturePath}. ` +
        "Run: python scripts/generate_contract_vectors.py",
    );
  }
  return JSON.parse(fs.readFileSync(fixturePath, "utf-8")) as ContractVectors;
}

/** Strip fixture metadata keys (prefixed with _) before parsing. */
function stripMeta(obj: Record<string, unknown>): Record<string, unknown> {
  return Object.fromEntries(
    Object.entries(obj).filter(([k]) => !k.startsWith("_")),
  );
}

// Load once at module level to avoid re-reading per test.
const fixture = loadFixture();

// ---------------------------------------------------------------------------
// Valid vector tests
// ---------------------------------------------------------------------------

describe("RiskScoreSchema – valid contract vectors", () => {
  it("parses the 'complete' vector with all fields including latency_ms", () => {
    const raw = stripMeta(fixture.risk_score.complete);
    const result = RiskScoreSchema.safeParse(raw);
    expect(result.success, `Parse failed: ${JSON.stringify((result as { error?: unknown }).error)}`).toBe(true);

    if (!result.success) return;
    const score = result.data;

    expect(score.wallet).toBeTruthy();
    expect(score.asset_pair).toBeTruthy();
    expect(score.score).toBeGreaterThanOrEqual(0);
    expect(score.score).toBeLessThanOrEqual(100);
    expect(typeof score.benford_flag).toBe("boolean");
    expect(typeof score.ml_flag).toBe("boolean");
    expect(score.confidence).toBeGreaterThanOrEqual(0);
    expect(score.confidence).toBeLessThanOrEqual(100);

    // latency_ms: must be present and a number in the complete vector
    expect(score.latency_ms).not.toBeNull();
    expect(score.latency_ms).not.toBeUndefined();
    expect(typeof score.latency_ms).toBe("number");
  });

  it("parses the 'minimal' vector with all optional fields null", () => {
    const raw = stripMeta(fixture.risk_score.minimal);
    const result = RiskScoreSchema.safeParse(raw);
    expect(result.success).toBe(true);

    if (!result.success) return;
    expect(result.data.latency_ms ?? null).toBeNull();
    expect(result.data.score_lower ?? null).toBeNull();
    expect(result.data.score_upper ?? null).toBeNull();
    expect(result.data.prediction_set ?? null).toBeNull();
    expect(result.data.coverage_guarantee ?? null).toBeNull();
  });

  it("parses the 'disputed' vector with disputed=true", () => {
    const raw = stripMeta(fixture.risk_score.disputed);
    const result = RiskScoreSchema.safeParse(raw);
    expect(result.success).toBe(true);
    if (result.success) {
      expect(result.data.disputed).toBe(true);
    }
  });

  it("parses 'score_boundary_zero' (score=0)", () => {
    const raw = stripMeta(fixture.risk_score.score_boundary_zero);
    const result = RiskScoreSchema.safeParse(raw);
    expect(result.success).toBe(true);
    if (result.success) {
      expect(result.data.score).toBe(0);
    }
  });

  it("parses 'score_boundary_hundred' (score=100)", () => {
    const raw = stripMeta(fixture.risk_score.score_boundary_hundred);
    const result = RiskScoreSchema.safeParse(raw);
    expect(result.success).toBe(true);
    if (result.success) {
      expect(result.data.score).toBe(100);
    }
  });

  it("parses v2+ uncertainty fields correctly in 'complete' vector", () => {
    const raw = stripMeta(fixture.risk_score.complete);
    const result = RiskScoreSchema.safeParse(raw);
    expect(result.success).toBe(true);
    if (!result.success) return;

    expect(result.data.score_lower).not.toBeNull();
    expect(result.data.score_upper).not.toBeNull();
    expect(result.data.prediction_set).not.toBeNull();
    expect(result.data.coverage_guarantee).not.toBeNull();
    expect(Array.isArray(result.data.prediction_set)).toBe(true);
  });

  it("all valid vector keys deserialize without error", () => {
    const validKeys = [
      "complete",
      "minimal",
      "disputed",
      "score_boundary_zero",
      "score_boundary_hundred",
    ];
    for (const key of validKeys) {
      const raw = stripMeta(fixture.risk_score[key]);
      const result = RiskScoreSchema.safeParse(raw);
      expect(
        result.success,
        `Vector '${key}' failed: ${JSON.stringify((result as { error?: unknown }).error)}`,
      ).toBe(true);
    }
  });
});

// ---------------------------------------------------------------------------
// Adversarial vector tests (divergence detection)
// ---------------------------------------------------------------------------

describe("RiskScoreSchema – adversarial vectors (divergence detection)", () => {
  it("rejects a payload with 'risk_score' instead of 'score' (wrong field name)", () => {
    /**
     * This test proves divergence detection: a field rename in the Python model
     * that is not reflected in the TS schema will cause a parse failure here,
     * provided the renamed field is also absent from the Zod schema.
     *
     * The adversarial vector uses 'risk_score' instead of 'score'. Since 'score'
     * is required in RiskScoreSchema, the parse must fail with a ZodError.
     */
    const raw = stripMeta(fixture.risk_score_adversarial.wrong_field_name);
    const result = RiskScoreSchema.safeParse(raw);
    expect(result.success).toBe(false);
    if (!result.success) {
      const err = result.error as ZodError;
      const fieldPaths = err.errors.map((e) => e.path.join("."));
      expect(fieldPaths).toContain("score");
    }
  });

  it("rejects a payload with score=999 (out of range)", () => {
    const raw = stripMeta(fixture.risk_score_adversarial.score_out_of_range);
    const result = RiskScoreSchema.safeParse(raw);
    expect(result.success).toBe(false);
    if (!result.success) {
      const err = result.error as ZodError;
      const fieldPaths = err.errors.map((e) => e.path.join("."));
      expect(fieldPaths).toContain("score");
    }
  });

  it("rejects a payload where 'score' is renamed to 'SCORE' (case drift)", () => {
    /**
     * Simulate what happens when a language implementation uses camelCase or
     * wrong casing. This is the canonical cross-language drift scenario.
     */
    const raw = stripMeta(fixture.risk_score.complete);
    const drifted = { ...raw, SCORE: raw.score };
    delete (drifted as Record<string, unknown>).score;

    const result = RiskScoreSchema.safeParse(drifted);
    expect(result.success).toBe(false);
    if (!result.success) {
      const err = result.error as ZodError;
      const fieldPaths = err.errors.map((e) => e.path.join("."));
      expect(fieldPaths).toContain("score");
    }
  });
});

// ---------------------------------------------------------------------------
// Required fields agreement check
// ---------------------------------------------------------------------------

describe("RiskScoreSchema – required fields agreement with fixture", () => {
  it("RiskScoreSchema includes all fields declared in required_risk_score_fields", () => {
    /**
     * This test checks that every field the Python canonical model declares as
     * required is actually present in the Zod schema. If a new field is added to
     * the Python model and not to the Zod schema, the round-trip test for the
     * 'complete' vector will catch it.
     *
     * We check by validating the 'complete' vector (which has all fields populated)
     * and confirming the parsed output contains all required fields.
     */
    const raw = stripMeta(fixture.risk_score.complete);
    const result = RiskScoreSchema.safeParse(raw);
    expect(result.success).toBe(true);
    if (!result.success) return;

    const parsedKeys = new Set(Object.keys(result.data));

    // Check each required field is in the parsed output
    // (optional fields may not appear if null, but should be parseable)
    const alwaysPresent = [
      "wallet",
      "asset_pair",
      "score",
      "benford_flag",
      "ml_flag",
      "confidence",
      "disputed",
      "timestamp",
    ];
    for (const field of alwaysPresent) {
      expect(parsedKeys.has(field), `Required field '${field}' absent from parsed output`).toBe(
        true,
      );
    }
  });

  it("latency_ms is in the Zod schema", () => {
    /**
     * Explicit check for the canonical divergence example: latency_ms was absent
     * from the TypeScript SDK while present in the Python model.
     */
    // Parse a vector that has latency_ms populated
    const raw = stripMeta(fixture.risk_score.complete);
    const result = RiskScoreSchema.safeParse(raw);
    expect(result.success).toBe(true);
    if (!result.success) return;

    // If latency_ms were not in the schema, it would be stripped by Zod's default
    // strip behaviour and would be undefined/missing from the result.
    // We check that the raw value round-tripped.
    expect("latency_ms" in result.data).toBe(true);
    expect(result.data.latency_ms).not.toBeUndefined();
  });

  it("v2+ uncertainty fields are in the Zod schema", () => {
    /**
     * Acceptance criterion: 'The schema-contract-enforcement mechanism covers,
     * at minimum, RiskScore's uncertainty-quantification fields.'
     */
    const requiredV2Fields = [
      "score_lower",
      "score_upper",
      "prediction_set",
      "coverage_guarantee",
    ];
    const raw = stripMeta(fixture.risk_score.complete);
    const result = RiskScoreSchema.safeParse(raw);
    expect(result.success).toBe(true);
    if (!result.success) return;

    for (const field of requiredV2Fields) {
      expect(
        field in result.data,
        `v2+ field '${field}' absent from Zod schema output`,
      ).toBe(true);
    }
  });

  it("fixture _contract_version is present", () => {
    expect(fixture._contract_version).toBeTruthy();
  });
});

File: crates/ledgerlens-sdk/tests/contract_vectors_test.rs
/// Cross-language schema contract vector tests (Rust side).
///
/// These tests load the canonical fixture file `tests/fixtures/contract_vectors.json`
/// and verify that every valid vector deserializes cleanly via serde_json into the
/// `ledgerlens_sdk::models::RiskScore` struct. They also verify that adversarial
/// vectors (wrong field names, out-of-range values) are handled correctly.
///
/// ADR reference: docs/adr/ADR-005-schema-contract-enforcement.md
///
/// The fixture path is resolved relative to the workspace root using the
/// CARGO_MANIFEST_DIR environment variable set by cargo test.
use ledgerlens_sdk::models::RiskScore;
use serde_json::Value;
use std::fs;
use std::path::PathBuf;

/// Returns the path to the contract vectors fixture file.
///
/// During `cargo test` the current directory is the workspace root, so we
/// can resolve the fixture path relative to CARGO_MANIFEST_DIR's parent (the
/// workspace root).
fn fixture_path() -> PathBuf {
    let manifest_dir = std::env::var("CARGO_MANIFEST_DIR")
        .expect("CARGO_MANIFEST_DIR must be set by cargo test");
    // crates/ledgerlens-sdk → go up two levels to workspace root
    let workspace_root = PathBuf::from(manifest_dir)
        .parent()
        .expect("crate has parent dir")
        .parent()
        .expect("crates/ has parent dir")
        .to_path_buf();
    workspace_root
        .join("tests")
        .join("fixtures")
        .join("contract_vectors.json")
}

fn load_fixture() -> Value {
    let path = fixture_path();
    let content = fs::read_to_string(&path).unwrap_or_else(|e| {
        panic!(
            "Failed to read contract vectors fixture at {}: {}. \
             Run: python scripts/generate_contract_vectors.py",
            path.display(),
            e
        )
    });
    serde_json::from_str(&content).expect("contract_vectors.json must be valid JSON")
}

fn parse_risk_score(fixture: &Value, key: &str) -> RiskScore {
    let raw = &fixture["risk_score"][key];
    serde_json::from_value(raw.clone()).unwrap_or_else(|e| {
        panic!(
            "Failed to deserialize risk_score['{}'] into RiskScore: {}\nRaw: {}",
            key,
            e,
            serde_json::to_string_pretty(raw).unwrap_or_default()
        )
    })
}

// ---------------------------------------------------------------------------
// Valid vector tests
// ---------------------------------------------------------------------------

#[test]
fn contract_vector_risk_score_complete() {
    let fixture = load_fixture();
    let score = parse_risk_score(&fixture, "complete");

    assert!(!score.wallet.is_empty(), "wallet must not be empty");
    assert!(!score.asset_pair.is_empty(), "asset_pair must not be empty");
    assert!(score.score <= 100, "score must be in 0-100 range");
    assert!(score.confidence <= 100, "confidence must be in 0-100 range");

    // latency_ms must be present in the complete vector
    assert!(
        score.latency_ms.is_some(),
        "latency_ms is None in 'complete' vector but should be populated. \
         This field exists in the canonical Python model (detection/risk_score.py)."
    );

    // v2+ uncertainty fields must be present in the complete vector
    assert!(
        score.score_lower.is_some(),
        "score_lower missing from complete vector (v2+ uncertainty field)"
    );
    assert!(
        score.score_upper.is_some(),
        "score_upper missing from complete vector (v2+ uncertainty field)"
    );
    assert!(
        score.prediction_set.is_some(),
        "prediction_set missing from complete vector (v2+ uncertainty field)"
    );
    assert!(
        score.coverage_guarantee.is_some(),
        "coverage_guarantee missing from complete vector (v2+ uncertainty field)"
    );
}

#[test]
fn contract_vector_risk_score_minimal() {
    let fixture = load_fixture();
    let score = parse_risk_score(&fixture, "minimal");

    // Optional fields must be None in the minimal vector
    assert!(score.latency_ms.is_none(), "latency_ms should be None in minimal vector");
    assert!(score.score_lower.is_none());
    assert!(score.score_upper.is_none());
    assert!(score.prediction_set.is_none());
    assert!(score.coverage_guarantee.is_none());
}

#[test]
fn contract_vector_risk_score_disputed() {
    let fixture = load_fixture();
    let score = parse_risk_score(&fixture, "disputed");
    assert!(score.disputed, "disputed must be true in the disputed vector");
}

#[test]
fn contract_vector_risk_score_boundary_zero() {
    let fixture = load_fixture();
    let score = parse_risk_score(&fixture, "score_boundary_zero");
    assert_eq!(score.score, 0, "score must be 0 (boundary value)");
}

#[test]
fn contract_vector_risk_score_boundary_hundred() {
    let fixture = load_fixture();
    let score = parse_risk_score(&fixture, "score_boundary_hundred");
    assert_eq!(score.score, 100, "score must be 100 (boundary value)");
}

// ---------------------------------------------------------------------------
// prediction_set type correctness
// ---------------------------------------------------------------------------

#[test]
fn contract_vector_prediction_set_is_vec_i32() {
    /// Verify that prediction_set elements are i32 (signed integer), not u8.
    ///
    /// The Python canonical type is `list[int]` (signed integers in the range
    /// of class indices, typically 0 and 1). A previous Rust implementation
    /// incorrectly used `Vec<u8>`, which would silently truncate or reject
    /// negative indices and diverge from the Python contract.
    let fixture = load_fixture();
    let score = parse_risk_score(&fixture, "complete");

    if let Some(ps) = &score.prediction_set {
        // This assertion is primarily a type-system check (the Rust type
        // already enforces Vec<i32>), but we also verify the value round-trips.
        assert!(
            !ps.is_empty(),
            "prediction_set in 'complete' vector should not be empty"
        );
        // Verify that class index 1 (wash-trading class) round-trips correctly.
        // If Vec<u8> were still used, negative indices would fail to deserialize.
        for &idx in ps {
            assert!(
                idx >= 0,
                "prediction_set index {} is negative — class indices start at 0",
                idx
            );
        }
    }
}

// ---------------------------------------------------------------------------
// Adversarial vector tests (divergence detection)
// ---------------------------------------------------------------------------

#[test]
fn contract_vector_adversarial_wrong_field_name_rejected() {
    /// A payload with 'risk_score' instead of 'score' must fail deserialization.
    ///
    /// serde_json with #[serde(deny_unknown_fields)] would reject this outright.
    /// Without that attribute, the missing 'score' field (required) causes a
    /// missing-field error. Either outcome proves the deserializer is not silently
    /// accepting the renamed field.
    ///
    /// This test proves divergence detection: a field rename in Python that is
    /// not reflected in the Rust model will be caught here.
    let fixture = load_fixture();
    let raw = &fixture["risk_score_adversarial"]["wrong_field_name"];

    let result: Result<RiskScore, _> = serde_json::from_value(raw.clone());

    // The deserialization must fail because 'score' is missing (renamed to 'risk_score').
    // RiskScore.score is a required field with no default.
    assert!(
        result.is_err(),
        "Adversarial vector with 'risk_score' instead of 'score' was unexpectedly accepted. \
         This means a field rename in the Python model would go undetected in Rust."
    );
}

#[test]
fn contract_vector_all_valid_vectors_deserialize() {
    /// Meta-test: every non-adversarial RiskScore vector must deserialize without error.
    let fixture = load_fixture();
    let valid_keys = ["complete", "minimal", "disputed", "score_boundary_zero", "score_boundary_hundred"];

    for key in &valid_keys {
        let raw = &fixture["risk_score"][key];
        let result: Result<RiskScore, _> = serde_json::from_value(raw.clone());
        assert!(
            result.is_ok(),
            "Valid vector '{}' failed to deserialize: {:?}",
            key,
            result.err()
        );
    }
}

// ---------------------------------------------------------------------------
// Required fields agreement check
// ---------------------------------------------------------------------------

#[test]
fn contract_vector_required_fields_declared_in_fixture() {
    /// Verify the fixture declares all expected required fields for RiskScore.
    ///
    /// This acts as the 'agreement check': if a field is added to the Rust struct
    /// that is not in the fixture, this test prompts the developer to update the
    /// fixture (and then all language implementations).
    let fixture = load_fixture();
    let declared: Vec<String> = serde_json::from_value(
        fixture["required_risk_score_fields"].clone()
    )
    .expect("required_risk_score_fields must be a JSON array of strings");

    // These are the fields the Rust RiskScore struct knows about.
    // If a new field is added here that's not in the fixture, this test fails.
    let rust_fields = [
        "wallet", "asset_pair", "score", "benford_flag", "ml_flag",
        "confidence", "disputed", "timestamp", "latency_ms",
        "score_lower", "score_upper", "prediction_set", "coverage_guarantee",
    ];

    for field in &rust_fields {
        assert!(
            declared.iter().any(|d| d == field),
            "Rust struct field '{}' is not declared in fixture's required_risk_score_fields. \
             Update the fixture and all language implementations. \
             Run: python scripts/generate_contract_vectors.py",
            field
        );
    }
}
